src/t8_schemes/t8_quads_w_rectangular_subelements/t8_subelements/t8_subelements_quad_cxx.hxx

```diff
@@ -0,0 +1,414 @@
+ *     subelement_id (what subelement of the transition cell is the given subelement?)
+ * 
+ * In order to refine a quad element using subelements, it is important to know these additional information. 
+ * We can use them for example to determine the coordinates of the subelement vertices, 
+ * since they can differ for subelements of the same id but a different subelement type. */
+
+#ifndef T8_SUBELEMENTS_QUAD_CXX_HXX
+#define T8_SUBELEMENTS_QUAD_CXX_HXX
+
+#include <p4est.h>
+#include <t8_element_cxx.hxx>
+
+#include "t8_schemes/t8_default/t8_default_line_cxx.hxx"
+#include "t8_schemes/t8_default/t8_default_common_cxx.hxx"
+
+/** The structure holding a quadrilateral element in the default scheme.
+ * We make this definition public for interoperability of element classes.
+ * We might want to put this into a private, scheme-specific header file.
+ */
+
+/* Define the struct, that stores all information needed for the quad scheme and subelements.
+ * 
+ *         p4est quadrant        recursive quad       refinement, using a transition 
+ *                                 refinement             cell with subelements
+ *         x - - - - - x         x - - x - - x                x - - - - - x
+ *         |           |         |     |     |                | \   2   / |       
+ *         |           |         |     |     |                | 1 \   /   |
+ *         |           |   -->   x - - x - - x       or       x - - x   3 |
+ *         |           |         |     |     |                | 0 / | \   |
+ *         |           |         |     |     |                | / 5 | 4 \ |
+ *         x - - - - - x         x - - x - - x                x - - x - - x 
+ * 
+ * A p4est quadrant can be refined, using either the standard quad scheme, or a transition cell, consisting of different subelements. 
+ * The quad refinement scheme is recursive, whereas a transition cell can only be used once, for example to remove hanging nodes, after the mesh has been adapted and balanced. 
+ * There are different types of transition cells possible, which we will refer to as subelement_type. 
+ * Each transition cell consists of different subelements. The given example consists of 6 different subelements, whose ids range from 0 to 5.
+ * A dummy variable will store the information, whether a given element is a subelement or a standard quad element. */
+
+typedef struct
+{
+  p4est_quadrant_t    p4q;      /* p4est quadrant */
+  int                 dummy_is_subelement;      /* saves the information, whether an element is a subelement (1: is subelement, -1: is no subelement) */
+  int                 subelement_type;  /* saves the information, which type of transition cell a subelement is associated to */
+  int                 subelement_id;    /* saves the information, what children subelement the given element is */
+} t8_quad_with_subelements;
+
+typedef t8_quad_with_subelements t8_pquad_t;
+
+/** define some subelement related constants */
```

Comment:
nice
id suggest to make the names more quad specific, like
T8_SUB_QUAD_MAX_SUBELEMENT_TYPE

Reply:
done